### ISSUE TITLE
provider/aws: Allow specifying credentials provider

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform#1822]
Credentials are currently detected using multiple providers in certain order.
This implies a few things:
- not everyone may like the order of detection
- sometimes people may just explicitly want to use one particular provider and force Terraform to use that instead of doing any detection
- anyone reading TF code may not even realise what will happen (loading keys from certain locations) - imagine a situation where I've got production env keys set in `default` profile in `~/.aws/credentials` and forget to define access_key & secret key, then run TF in production ENV by accident. `provider.aws.allowed_account_ids` which I added recently should not IMO be the only way to prevent this from happening
- I'd expect to be able to use TF across different ENVs with different keys without having to touch `~/.aws/credentials` and set certain profile as default

This is to extract the original idea behind https://github.com/hashicorp/terraform/pull/1049 which then turned into somewhat much bigger PR solving too many problems.

**Comment by phinze** [comment]
Hi @radeksimko - is this still a valid issue? The AWS authentication code path has gone through multiple iterations since this was filed. Happy to keep it open if so, just checking! 😁

**Comment by tilgovi** [comment]
I don't fully understand the ask, but as far as I can tell setting credentials from the environment works fine. But maybe the ask is to error out if, say, `AWS_PROFILE` isn't set? It sounds like the ask is that @radeksimko wants to use environmental configuration but doesn't want to accidentally use the wrong account if it's the default. I don't really understand, though. In that case, can't you just have non-defaulted variables that the user is forced to provide and configure the provider with those?

**Comment by radeksimko** [comment]
I still think it would be nice to make this explicit, rather than implicit, so that I can say e.g. _"Always use ENV variables and never fall back to EC2 metadata API"_.

I think I could make this possible fairly easily after https://github.com/hashicorp/terraform/pull/5030 which also includes a decent refactoring. I'm happy to come up with another PR after #5030 .

> I'd expect to be able to use TF across different ENVs with different keys without having to touch ~/.aws/credentials and set certain profile as default

This is now possible, although not for users that use S3 remote state. Remote state can be fixed via https://github.com/hashicorp/terraform/pull/5270 (ideally after #5030 )

**Comment by apparentlymart** [comment]
Hi everyone,

While there are certainly some good use-cases for this feature, we (the Terraform team) don't have any short-term plans to implement this ourselves. We're trying to prune stale feature requests (that aren't likely to be addressed soon) by closing them.

If anyone is still interested in having Terraform support this, we would gladly review a pull request.
